File: datastructures/linkedlist/java/impl/DoublyLinkedList.java
public class DoublyLinkedList<T> {

    private int size;
    private Node<T> head;

    public DoublyLinkedList() {
        head = null;
    }

    public T get(int index) {
        if (index > size) {
            return null;
        }

        Node<T> current = head;

        for (int i = 0; i < index; i++) {
            current = current.next;
        }

        if (current != null) {
            return current.data;
        } else {
            return null;
        }
    }

    public void addAtHead(T val) {
        if (head == null) {
            head = new Node<T>(val);
        } else {
            Node<T> newNode = new Node<T>(val);
            Node<T> current = head;

            newNode.next = current;
            current.previous = newNode;
            head = newNode;
        }

        size++;
    }

    public void addAtTail(T val) {
        if (head == null) {
            head = new Node<T>(val);
        } else {
            Node<T> current = head;
            Node<T> newNode = new Node<T>(val);

            while (current.next != null) {
                current = current.next;
            }

            current.next = newNode;
            newNode.previous = current;
        }

        size++;
    }

    public void addAtIndex(int index, T val) {
        if (head == null) {
            head = new Node<T>(val);
        } else {
            if (index == 0) {
                addAtHead(val);
            } else if (index >= size) {
                addAtTail(val);
            } else {
                Node<T> current = head;
                Node<T> newNode = new Node<T>(val);

                for (int i = 0; i < index - 1; i++) {
                    current = current.next;
                }

                Node<T> next = current.next;
                newNode.previous = current;
                newNode.next = next;
                current.next = newNode;
                next.previous = newNode;

                size++;
            }
        }
    }

    public void deleteAtIndex(int index) {
        if (head == null || index >= size) {
            return;
        }

        if (index == 0) {
            head = head.next;
            size--;
            return;
        }

        Node<T> current = head;

        for (int i = 0; i < index - 1; i++) {
            current = current.next;
        }

        Node<T> toDelete = current.next;
        current.next = toDelete.next;
        if (toDelete.next != null) {
            toDelete.next.previous = current;
        }

        size--;
    }

    public int size() {
        return size;
    }

    public void print() {
        Node<T> current = head;
        while (current != null) {
            System.out.print(current.data.toString() + " ");
            current = current.next;
        }
    }

    private static class Node<T> {
        private T data;
        private Node<T> next;
        private Node<T> previous;

        public Node(T data) {
            this.data = data;
        }

        @Override
        public String toString() {
            return data.toString();
        }
    }
}

File: datastructures/linkedlist/python/doubly-linked-list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.previous = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    def get(self, index):
        if not self.head:
            return -1

        current = self.head
        for _ in range(index):
            current = current.next

        if current:
            return current.value
        else:
            return -1

    def add_at_head(self, val):
        if not self.head:
            self.head = Node(val)
        else:
            current = self.head
            new_node = Node(val)
            new_node.next = current
            current.previous = new_node
            self.head = new_node

        self.size += 1

    def add_at_tail(self, val):
        if not self.head:
            self.head = Node(val)
        else:
            current = self.head
            while current.next:
                current = current.next
            new_node = Node(val)
            current.next = new_node
            new_node.previous = current

        self.size += 1

    def add_at_index(self, index, val):
        if not self.head:
            self.head = Node(val)
        else:
            if index == 0:
                self.add_at_head(val)
            else:
                current = self.head
                for _ in range(index - 1):
                    current = current.next
                new_node = Node(val)
                new_node.next = current.next
                current.next = new_node
                new_node.prev = current

                self.size += 1

    def delete_at_index(self, index):
        if self.head and index < self.size:
            current = self.head
            if index == 0:
                self.head = self.head.next
            else:
                for _ in range(index - 1):
                    current = current.next
                next = current.next.next
                current.next = next
                if next:
                    next.previous = current

            self.size -= 1

    def print(self):
        current = self.head
        while current is not None:
            print(current.value)
            current = current.next


def main():
    doubly_linked_list = DoublyLinkedList()

    doubly_linked_list.add_at_tail(7)

    doubly_linked_list.print()
    # print(doubly_linked_list.get(0))


if __name__ == "__main__":
    main()
